src/warc2zim/icon_finder.py

```diff
@@ -0,0 +1,132 @@
+from collections.abc import Iterable
+from dataclasses import dataclass
+
+from bs4 import BeautifulSoup, Tag
+
+from warc2zim.constants import logger
+
+ZIM_ILLUSTRATION_SIZE = 48
+
+
+@dataclass(frozen=True)
+class IconUrlAndSize:
+    """Helper class to compare icons for warc2zim usage"""
+
+    url: str
+    width: int | None
+    height: int | None
+
+    def __lt__(self, other):
+        return IconUrlAndSize._cmp_icons(self, other) < 0
+
+    def __gt__(self, other):
+        return IconUrlAndSize._cmp_icons(self, other) > 0
+
+    @classmethod
+    def _cmp_icons(cls, x: "IconUrlAndSize", y: "IconUrlAndSize") -> int:
+        if x.width and y.width and x.height and y.height:
+            # icons are supposed to be squared ; should they not be squared, we consider
+            # only the smallest dimension for comparison
+            x_dim = min(x.width, x.height)
+            y_dim = min(y.width, y.height)
+            if x_dim == ZIM_ILLUSTRATION_SIZE:
+                if y_dim != ZIM_ILLUSTRATION_SIZE:
+                    return 1  # x is perfect, y not
+                return 0  # both images are perfect
+            if x_dim > ZIM_ILLUSTRATION_SIZE:
+                if y_dim == ZIM_ILLUSTRATION_SIZE:
+                    return -1  # y is perfect, x not
+                elif x_dim < y_dim:
+                    return -1  # perfer biggest icon (y)
+                elif x_dim == y_dim:
+                    return 0  # both icon are equivalent
+                else:
+                    return 1  #  perfer biggest icon (x)
+            if y_dim >= ZIM_ILLUSTRATION_SIZE:  # and x < ZIM_ILLUSTRATION_SIZE
+                return -1  # prefer biggest icon (y)
+            elif (
+                x_dim < y_dim
+            ):  #  and x < ZIM_ILLUSTRATION_SIZE and y < ZIM_ILLUSTRATION_SIZE
+                return -1  # prefer biggest icon (y)
+            elif (
+                x_dim == y_dim
+            ):  #  and x < ZIM_ILLUSTRATION_SIZE and y < ZIM_ILLUSTRATION_SIZE
+                return 0  # both are too small
+            else:
+                return 1  # prefer biggest icon (x)
+        elif x.width and x.height and not y.width:
+            x_dim = min(x.width, x.height)
+            if x_dim >= ZIM_ILLUSTRATION_SIZE and x_dim >= ZIM_ILLUSTRATION_SIZE:
+                return 1  # prefer x which is known to be big enough
+            else:
+                return 0  # x has height but it is too small, y we don't know
+        elif not x.width and y.width and y.height:
+            y_dim = min(y.width, y.height)
+            if y_dim >= ZIM_ILLUSTRATION_SIZE and y_dim >= ZIM_ILLUSTRATION_SIZE:
+                return -1  # prefer y which is known to be big enough
+            else:
+                return 0  # y has height but it is too small, x we don't know
+        else:  # all other cases where we miss some sizes in x and y
+            return 0
+
+
+def sort_icons(icons: Iterable[IconUrlAndSize]) -> list[IconUrlAndSize]:
```

Comment:
This is misleading. It's not sorting the iterable, it's returning a sorted version. Given stdlib already has the sorted (return) and sort (in place) concepts, we should stick to it.

```suggestion
def get_sorted_icons(icons: Iterable[IconUrlAndSize]) -> list[IconUrlAndSize]:
```

Reply:
Agreed